Fall back to the legacy file/folder dialog on COMException

kpreisser opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Fixes #2506


## Proposed changes

- When creating an instance of the `NativeFileOpenDialog` or `NativeFileSaveDialog` fails with a `COMException`, fall back to the legacy file dialog instead of letting the exception bubble up to the caller.

Note: I'm not sure if there's another way to check if the COM object will be available without actually throwing (and catching) a `COMException`.

<!-- We are in TELL-MODE the following section must be completed -->

## Customer Impact

- When using the [`OpenFileDialog`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.openfiledialog?view=netcore-3.1), [`SaveFileDialog`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.savefiledialog?view=netcore-3.1), or [`FolderBrowserDialog`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.folderbrowserdialog?view=netcore-3.1) (with **`AutoUpgradeEnabled=true`**) on a Windows system lacking the necessary COM components (e.g. Windows Server Core), customers will experience the legacy file/folder dialog being shown when calling `ShowDialog()`, rather than a `COMException` being thrown preventing the dialog from working.

## Regression? 

- No

## Risk

-

<!-- end TELL-MODE -->


## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![grafik](https://user-images.githubusercontent.com/13289184/71029761-58216b80-2110-11ea-8b27-0e4f1f38657a.png)


### After

![grafik](https://user-images.githubusercontent.com/13289184/71029872-99198000-2110-11ea-8ae7-de4783a9d5cf.png)

![grafik](https://user-images.githubusercontent.com/13289184/71029904-a6cf0580-2110-11ea-8807-607ddd2d6e10.png)


## Test methodology <!-- How did you ensure quality? -->

- Manual testing on Windows Server 2019 (installed as Server Core) as described in #2506, as well as on Windows 10 to verify that the Vista dialogs are still shown

<!--
     Microsoft prioritizes making our products accessible. 
     WinForms has a key role in allowing developers to create accessible apps. 
     
     When submitting a change which impacts UI in any way, including adding new UI or
     modifying existing controls the developer needs to run the Accessibility Insights
     tool (https://accessibilityinsights.io/) and verify that there are no changes or
     regressions. 
     
     The developer should run the Fast Pass over the impacted control(s) and provide
     a snapshot of the passing results along with before/after snapshots of the UI.
     More info: (https://accessibilityinsights.io/docs/en/web/getstarted/fastpass)
  -->


 

## Test environment(s) <!-- Remove any that don't apply -->

.NET Core SDK (reflecting any global.json):
 Version:   5.0.100-alpha1-015777
 Commit:    cea32db3f2

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100-alpha1-015777\

Host (useful for support):
  Version: 5.0.0-alpha.1.19564.1
  Commit:  c77948d92a

.NET Core SDKs installed:
  3.1.100 [C:\Program Files\dotnet\sdk]
  5.0.100-alpha1-015777 [C:\Program Files\dotnet\sdk]


<!-- Mention language, UI scaling, or anything else that might be relevant -->


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2515)

RussKie commented on src/System.Windows.Forms/src/System/Windows/Forms/FileDialog_Vista.cs:
```diff
@@ -32,15 +33,29 @@ internal bool UseVistaDialogInternal
+            {
+                // Creating the Vista dialog can fail on Windows Server Core, even if the
+                // Server Core App Compatibility FOD is installed.
+                // See: https://github.com/dotnet/winforms/issues/2506
```
Please remove this line, this change can be traced back through git log to the underlying issue.
The commit message also references it.

kpreisser replied:
Done.

kpreisser commented:
I have removed the link to the Github issue, and added a paragraph to the commit message explaining the `COMException` on Windows Server Core.

Thanks!

RussKie commented:
Thank you
